server/src/files.rs

```diff
@@ -0,0 +1,137 @@
+use std::{fmt, path::PathBuf, time::Duration};
+
+use opendal::{services::S3, Operator};
+use tokio::fs::File;
+
+use crate::{appstate::AppState, config::Config, errors::AtomicServerResult};
+
+pub enum FileStore {
+    S3,
+    FS,
+}
+
+impl FileStore {
+    const S3_PREFIX: &'static str = "s3:";
+    const FS_PREFIX: &'static str = "fs:";
+
+    pub fn get_config_file_store(config: &Config) -> FileStore {
+        if config.opts.s3_bucket.is_some() {
+            FileStore::S3
+        } else {
+            FileStore::FS
+        }
+    }
+
+    pub fn get_subject_file_store(subject: &str) -> FileStore {
+        if subject.contains(Self::S3_PREFIX) {
+            FileStore::S3
+        } else {
+            FileStore::FS
+        }
+    }
+
+    pub fn prefix(&self) -> &str {
+        match self {
+            Self::S3 => Self::S3_PREFIX,
+            Self::FS => Self::FS_PREFIX,
+        }
+    }
+
+    pub fn encoded(&self) -> String {
+        urlencoding::encode(self.prefix()).into_owned()
+    }
+}
+
+impl fmt::Display for FileStore {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.prefix())
+    }
+}
+
+pub async fn s3_upload_object(
+    appstate: &AppState,
+    file_id: &str,
+    file_path: &PathBuf,
+) -> AtomicServerResult<()> {
+    let mut builder = S3::default();
+    builder.bucket(appstate.config.opts.s3_bucket.as_ref().unwrap().as_str());
```

Comment:
Unwrap can't go here, it would cause a crash at runtime

Reply:
Ah yes. I originally implemented this way because we check that the bucket exists before calling this function so I knew it'd exist here, but it definitely introduces some brittleness, especially since it's in its own module. Great call out.